distributed/worker.py

```diff
@@ -1663,52 +1634,52 @@ def transition_ready_executing(self, key):
+    def transition_ready_memory(self, ts, value=None):
+        self.send_task_state_to_scheduler(ts)
 
-    def transition_constrained_executing(self, key):
-        self.transition_ready_executing(key)
-        for resource, quantity in self.resource_restrictions[key].items():
+    def transition_constrained_executing(self, ts):
+        self.transition_ready_executing(ts)
+        for resource, quantity in ts.resource_restrictions.items():
             self.available_resources[resource] -= quantity
 
         if self.validate:
             assert all(v >= 0 for v in self.available_resources.values())
 
-    def transition_executing_done(self, key, value=no_value, report=True):
+    def transition_executing_done(self, ts, value=no_value, report=True):
         try:
             if self.validate:
-                assert key in self.executing or key in self.long_running
-                assert key not in self.waiting_for_data
-                assert key not in self.ready
+                assert not ts.waiting_for_data
+                assert ts.key not in self.ready
 
             out = None
-            if key in self.resource_restrictions:
-                for resource, quantity in self.resource_restrictions[key].items():
+            if ts.resource_restrictions is not None:
+                for resource, quantity in ts.resource_restrictions.items():
                     self.available_resources[resource] += quantity
 
-            if self.task_state[key] == "executing":
-                self.executing.remove(key)
+            if ts.state == "executing":
+                self.executing.discard(ts.key)
                 self.executed_count += 1
-            elif self.task_state[key] == "long-running":
-                self.long_running.remove(key)
+            elif ts.state == "long-running":
+                self.long_running.remove(ts.key)
 
             if value is not no_value:
                 try:
-                    self.task_state[key] = "memory"
-                    self.put_key_in_memory(key, value, transition=False)
+                    self.put_key_in_memory(ts, value, transition=False)
                 except Exception as e:
                     logger.info("Failed to put key in memory", exc_info=True)
                     msg = error_message(e)
-                    self.exceptions[key] = msg["exception"]
-                    self.tracebacks[key] = msg["traceback"]
-                    self.task_state[key] = "error"
+                    ts.exception = msg["exception"]
+                    ts.traceback = msg["traceback"]
+                    ts.state = "error"
                     out = "error"
 
-                if key in self.dep_state:
-                    self.transition_dep(key, "memory")
+                # Don't release the dependency keys, but do remove them from `dependents`
+                for dependency in ts.dependencies:
+                    dependency.dependents.discard(ts)
+                ts.dependencies = set()
```

Comment:
Tiny nit, but maybe `.clear()`?

If this was in the the scheduler I would recommend this due to what I would expect to be a small performance gain.  It probably doesn't matter as much here though.

Reply:
Sounds good.